### PR TITLE
Updating inbox source generators to Roslyn 4.4 and removing polyfill approach

### PR DESCRIPTION
**Pull request by joperezr**
This is replaying the changes that were made into https://github.com/dotnet/runtime/pull/74822 but on runtime's main branch. The plan is to get this into main branch, and then backport to rc2 branch since the change on release/7.0 branch had to be rolled back due to depending on RC1 shipping first.

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs [hunk 1]
```diff
@@ -27,7 +29,9 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
         {
             IncrementalValuesProvider<ClassDeclarationSyntax> classDeclarations = context.SyntaxProvider
                 .ForAttributeWithMetadataName(
+#if !ROSLYN4_4_OR_GREATER
                     context,
+#endif
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs:34]: why do we still polyfill here?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs:34]: This file is used in the generator that runs on downlevel versions of VS (so if someone uses System.Text.Json in a .NET Standard 2.0 project, they can still use the source-generated code in VS 17.x where `x<4` or with the .NET 6 SDK and get better perf than the V1 generator APIs provide)

Comment (CyrusNajmabadi) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs:34]: Cool, thanks! :)

File: .editorconfig [hunk 2]
```diff
@@ -94,7 +94,8 @@ dotnet_style_object_initializer = true:suggestion
 dotnet_style_collection_initializer = true:suggestion
 dotnet_style_explicit_tuple_names = true:suggestion
 dotnet_style_coalesce_expression = true:suggestion
-dotnet_style_null_propagation = true:suggestion
+dotnet_style_null_propagation = false:suggestion   # Turning off given new warnings that came with the new analyzers and we don't want to take the risk of changing those now
+dotnet_style_prefer_compound_assignment = false:suggestion
```
Comment (jkoritzinsky) [inline .editorconfig:98]: I think I have all of these fixed in the main branch, so I don't think we need to disable the rules here.

File: eng/Versions.props [hunk 3]
```diff
@@ -38,16 +38,16 @@
...
     <MicrosoftCodeAnalysisAnalyzersVersion>3.3.3</MicrosoftCodeAnalysisAnalyzersVersion>
-    <MicrosoftCodeAnalysisCSharpCodeStyleVersion>4.4.0-1.final</MicrosoftCodeAnalysisCSharpCodeStyleVersion>
-    <MicrosoftCodeAnalysisCSharpWorkspacesVersion>4.4.0-1.final</MicrosoftCodeAnalysisCSharpWorkspacesVersion>
-    <MicrosoftCodeAnalysisCSharpVersion>4.4.0-1.final</MicrosoftCodeAnalysisCSharpVersion>
+    <MicrosoftCodeAnalysisCSharpCodeStyleVersion>4.4.0-2.22423.18</MicrosoftCodeAnalysisCSharpCodeStyleVersion>
+    <MicrosoftCodeAnalysisCSharpWorkspacesVersion>4.4.0-2.22423.18</MicrosoftCodeAnalysisCSharpWorkspacesVersion>
+    <MicrosoftCodeAnalysisCSharpVersion>4.4.0-2.22423.18</MicrosoftCodeAnalysisCSharpVersion>
     <MicrosoftCodeAnalysisNetAnalyzersVersion>7.0.0-preview1.22452.2</MicrosoftCodeAnalysisNetAnalyzersVersion>
-    <MicrosoftCodeAnalysisVersion>4.4.0-1.final</MicrosoftCodeAnalysisVersion>
+    <MicrosoftCodeAnalysisVersion>4.4.0-2.22423.18</MicrosoftCodeAnalysisVersion>
```
Comment (jkoritzinsky) [inline eng/Versions.props:50]: @CyrusNajmabadi any chance you know when the `4.4.0-2.final` package will be published? We should be using the equivalent non-stable version here, but it's nice to use the stable one when possible.

**Review by jkoritzinsky — APPROVED** [review]

File: eng/Versions.props (hunk 3 above)
Comment (CyrusNajmabadi) [reply eng/Versions.props:50]: I am not the person to ask that.  Let's try @JoeRobich though!

Comment (joperezr) [reply eng/Versions.props:50]: FWIW, the version I picked is the one that shipped with RC1 SDK.

File: .editorconfig (hunk 2 above)
Comment (stephentoub) [reply .editorconfig:98]: We've generally also been controlling such analyzers via https://github.com/dotnet/runtime/blob/main/eng/CodeAnalysis.src.globalconfig.

**Review by stephentoub — APPROVED** [review]

**Comment by joperezr** [comment]
Failures are unrelated, so I'm merging now.

**Pull request merged**